core/drivers/crypto/caam/cipher/caam_cipher_xts.c

```diff
@@ -0,0 +1,254 @@
+	if (retstatus != CAAM_NO_ERROR) {
+		ret = TEE_ERROR_OUT_OF_MEMORY;
+		goto out;
+	}
+
+	retstatus = caam_cipher_block(ctx, false, NEED_KEY2, true, &ctx->tweak,
+				      &enc_tweak, false);
+	if (retstatus != CAAM_NO_ERROR) {
+		CIPHER_TRACE("Tweak encryption error");
+		ret = TEE_ERROR_GENERIC;
+		goto out;
+	}
+
+	/*
+	 * Encrypt or Decrypt input data.
+	 * Check if the last block is partial or not
+	 *  - if last block is partial, rebuild a complete
+	 *    block using the penultimate complete block
+	 *    encryption/decryption.
+	 *
+	 *  - else do all blocks.
+	 */
+
+	/* Calculate the number of complete block */
+	fullsize = dupdate->src.length;
+	lastblk = fullsize % ctx->alg->size_block;
+	fullsize -= lastblk;
+
+	/* One full block is needed */
+	if (!fullsize)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (lastblk)
+		fullsize -= ctx->alg->size_block;
+
+	srcbuf.data = dupdate->src.data;
+	srcbuf.length = ctx->alg->size_block;
+	srcbuf.paddr = psrc;
+
+	dstbuf.data = dupdate->dst.data;
+	dstbuf.length = ctx->alg->size_block;
+	dstbuf.paddr = pdst;
+
+	for (; fullsize > 0; fullsize -= ctx->alg->size_block) {
+		CIPHER_TRACE("Tweak block fullsize %zu", fullsize);
+		retstatus = do_tweak_block(ctx, &enc_tweak, &srcbuf, &dstbuf,
+					   &tmpdst);
+
+		CIPHER_TRACE("Tweak block ret 0x%" PRIx32, retstatus);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
+			goto out;
+		}
+
+		CIPHER_DUMPBUF("Source", srcbuf.data, srcbuf.length);
+		CIPHER_DUMPBUF("Dest", dstbuf.data, dstbuf.length);
+
+		/* Increment the source and destination block */
+		srcbuf.data += ctx->alg->size_block;
+		srcbuf.paddr += ctx->alg->size_block;
+
+		dstbuf.data += ctx->alg->size_block;
+		dstbuf.paddr += ctx->alg->size_block;
+	}
+
+	if (lastblk) {
+		CIPHER_TRACE("Last block size is %zu", lastblk);
+
+		/*
+		 * Allocate the temporary buffer containing the
+		 * penultimate block computed
+		 */
+		retstatus = caam_alloc_align_buf(&tmpsrc, ctx->alg->size_block);
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_OUT_OF_MEMORY;
+			goto out;
+		}
+
+		if (!ctx->encrypt) {
+			/*
+			 * In case of decryption, need to multiply
+			 * the tweak first
+			 */
+			memcpy(tmpsrc.data, enc_tweak.data, enc_tweak.length);
+			do_galois_mult(&tmpsrc);
+
+			retstatus = do_tweak_block(ctx, &tmpsrc, &srcbuf,
+						   &tmpdst, &tmpdst);
+		} else {
+			retstatus = do_tweak_block(ctx, &enc_tweak, &srcbuf,
+						   &tmpdst, &tmpdst);
+		}
+
+		CIPHER_TRACE("Tweak penultimate block ret 0x%" PRIx32,
+			     retstatus);
+
+		if (retstatus != CAAM_NO_ERROR) {
+			ret = TEE_ERROR_GENERIC;
+			goto out;
+		}
+
+		/* Build the last block and create the last destination block */
+		for (idx = 0; idx < lastblk; idx++) {
+			tmpsrc.data[idx] =
+				srcbuf.data[ctx->alg->size_block + idx];
+			dstbuf.data[ctx->alg->size_block + idx] =
+				tmpdst.data[idx];
+		}
+
+		for (; idx < ctx->alg->size_block; idx++)
+			tmpsrc.data[idx] = tmpdst.data[idx];
```

Comment:
`memcpy()`?

Reply:
We could do a `memcpy()` here indeed.
However doing like so might be better here :
A block size is 16 bytes and a part of the block has already been copied above (`lastblk`). `memcpy` copies byte per byte before being aligned on a 32 bits words and then speed up the copy.
calling `memcpy` implies pushing & poping stack.
That's why we prefer coping in a loop here

Reply:
ok

Reply:
So if I get it right lines 214 to 222 are optimizations with bytes copies in order to save us from calling memcpy() three times?

Perhaps some time should be spent on making the memcpy() calls faster. Depending on how the compiler optimizes it may very well unroll a memcpy() call inline. I haven't experimented much with yet since there hasn't been a need.

Reply:
> Depending on how the compiler optimizes it may very well unroll a memcpy() call inline.

True, and I suspect it could also do the opposite (replace an open-coded copy by a call to memcpy()). It's amazing what compilers are allowed to do ;-) for instance `printf("a");` will usually be replaced with `putchar('a');`